# nf-core/atacseq: Output

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

## Pipeline overview
The pipeline is built using [Nextflow](https://www.nextflow.io/). See [main README.md](../README.md) for a condensed listing of the pipeline, and the bioinformatics tools used at each step.

See [Illumina website](https://emea.illumina.com/science/sequencing-method-explorer/kits-and-arrays/atac-seq.html) for more information regarding the ATAC-seq protocol, and for an extensive list of publications.

The directories listed below will be created in the output directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Library-level analysis

The initial QC and alignments are performed at the library-level e.g. if the same library has been sequenced more than once to increase sequencing depth. This has the advantage of being able to assess each library individually, and the ability to process multiple libraries from the same sample in parallel.

1. **Raw read QC**

    *Software*:  
    [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/)  

    *Description*:  
    FastQC gives general quality metrics about your reads. It provides information about the quality score distribution across your reads, the per base sequence content (%T/A/G/C). You get information about adapter contamination and other overrepresented sequences. For further reading and documentation see the [FastQC help](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

    *Output directories*:
    * `fastqc/`  
      FastQC `*.html` files for read 1 (and read2 if paired-end) **before** adapter trimming.

    * `fastqc/zips/`    
      FastQC `*.zip` files for read 1 (and read2 if paired-end) **before** adapter trimming.    

2. **Adapter trimming**

    *Software*:  
    [Trim Galore!](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/)

    *Description*:  
    Trim Galore! is a wrapper tool around Cutadapt and FastQC to consistently apply quality and adapter trimming to FastQ files. By default, Trim Galore! will automatically detect and trim the appropriate adapter sequence. For most ATAC-seq datasets this will be the Nextera adapter sequence 'CTGTCTCTTATA'. See [usage.md](usage.md) for more details about the trimming options.

    *Output directories*:
    * `trim_galore/`  
      If `--saveTrimmed` is specified fastq files **after** adapter trimming will be placed in this directory.

    * `trim_galore/logs/`    
      Log files generated by Trim Galore!.

    * `trim_galore/fastqc/`    
      FastQC html files for read 1 (and read2 if paired-end) **after** adapter trimming.

    * `trim_galore/fastqc/zips/`    
      FastQC zip files for read 1 (and read2 if paired-end) **after** adapter trimming.

    *Plots*:  
    [MultiQC - Cutadapt trimmed sequence plot](images/mqc_cutadapt_plot.png)

3. **Alignment, duplicate marking and read filtering**

    *Software*:  
    [BWA](https://sourceforge.net/projects/bio-bwa/files/), [picard](https://broadinstitute.github.io/picard/), [SAMtools](https://sourceforge.net/projects/samtools/files/samtools/), [BEDTools](https://github.com/arq5x/bedtools2/), [BAMTools](https://github.com/pezmaster31/bamtools), [Pysam](http://pysam.readthedocs.io/en/latest/installation.html)

    *Description*:  
    Adapter-trimmed reads are mapped to the reference assembly using BWA. A genome index is required to run BWA so if this isnt provided explicitly using the `--bwa_index` parameter then it will be created automatically from the genome fasta input. The index creation process can take a while for larger genomes so it is possible to use the `--saveReference` parameter to save the indices for future pipeline runs, reducing processing times.

    Read duplicate marking is carried out using the Picard MarkDuplicates command. Duplicate reads are generally removed from the aligned reads to mitigate for fragments in the library that may have been sequenced more than once due to PCR biases. There is an option to keep duplicate reads with the `--keepDups` parameter but its generally recommend to remove them to avoid the wrong interpretation of the results. A similar option has been provided to keep reads that are multi-mapped - `--keepMultiMap`.

    Certain cell types and tissues yield an enormous fraction (typically 20â€“80%) of unusable sequences of mitochondrial origin. This is a known problem that is specific to ATAC-seq library preps - see [Montefiori et al. 2017](https://www.nature.com/articles/s41598-017-02547-w). There is an option to keep these reads using the `--keepMito` parameter but its generally recommended to remove these in order to get a more reliable assessment of the duplication rate from the rest of the genome.

    Other steps have been incorporated into the pipeline to clean the resulting alignments - see [main README.md](../README.md) for a more comprehensive listing and the tools used at each step.

    *Output directories*:
    * `bwa/library/`  
      Filtered, coordinate sorted alignment files in [BAM](https://samtools.github.io/hts-specs/SAMv1.pdf) format at the library-level. If `--saveAlignedIntermediates` is specified then the intermediate BAM files will be placed in this directory.

    * `bwa/library/flagstat/`    
      Multiple BAM files will be generated before the final filtered BAM file is created. The SAMtools flagstat files for a selection of these will be placed in this directory.

    * `bwa/library/idxstats/`    
      SAMtools idxstats files to determine the percentage of reads mapping to mitochondrial DNA.

    * `bwa/library/picard_metrics/`    
      Alignment QC files from picard CollectMultipleMetrics and the metrics file from MarkDuplicates.

    * `bwa/library/picard_metrics/pdf/`    
      Alignment QC plot files from picard CollectMultipleMetrics and the metrics file from MarkDuplicates.

    *Plots*:  
    [MultiQC - SAMtools idxstats plot](images/mqc_samtools_idxstats_plot.png)  
    [MultiQC - Picard deduplication stats plot](images/mqc_picard_deduplication_plot.png)  
    [MultiQC - Picard insert size plot](images/mqc_picard_insert_size_plot.png)

## Replicate-level analysis

The library-level alignments associated with any given sample are merged at the replicate-level.


1. **Alignment merging, duplicate marking and removal**

    *Software*:  
    [picard](https://broadinstitute.github.io/picard/), [SAMtools](https://sourceforge.net/projects/samtools/files/samtools/)

    *Description*:  
    TODO.

    *Output directories*:
    * `bwa/replicate/`  
      Replicate-level, merged, coordinate sorted BAM files after the re-marking and removal of duplicates.

    * `bwa/replicate/flagstat/`    
      Flagstat files associated with the final filtered merged BAM file.

    * `bwa/replicate/picard_metrics/`    
      Metrics file from MarkDuplicates.

2. **Normalised bigWig files**

    *Software*:  
    [BEDTools](https://github.com/arq5x/bedtools2/), [wigToBigWig](http://hgdownload.soe.ucsc.edu/admin/exe/)  

    *Description*:  
    TODO.

    *Output directories*:
    * `bwa/replicate/bigwig/`  
      Normalised [bigWig](https://genome.ucsc.edu/goldenpath/help/bigWig.html) files scaled to 1 million mapped reads.

3. **TSS meta-profiles**

    *Software*:  
    [deepTools](https://deeptools.readthedocs.io/en/develop/)  

    *Description*:  
    TODO.

    *Output directories*:
    * `bwa/replicate/deeptools/`  
      TSS meta-profile plot for coverage across all genes. Generated with deepTools *computeMatrix* and *plotProfile* commands.

    *Plots*:  
    [MultiQC - deepTools TSS meta-profile plot](images/mqc_deeptools_tss_plot.png)

4. **Call peaks**

    *Software*:  
    [MACS2](https://github.com/taoliu/MACS), [HOMER](http://homer.ucsd.edu/homer/download.html)

    *Description*:  
    TODO.

    *Output directories*:
    * `bwa/replicate/macs2/`  
      MACS2 output files: `*.xls`, `*.broadPeak` or `*.narrowPeak`, `*.gappedPeak` and `*summits.bed`. The files generated will depend on whether MACS2 has been run in narrowPeak or broadPeak mode.  
      HOMER peak-to-gene annotation file: `*.annotatePeaks.txt`.

    * `bwa/replicate/macs2/qc`  
      Peak QC plots including fold-change distribution and peak percentage across gene features: `*.pdf`.  
      MultiQC custom-content files for [FRiP score](https://genome.cshlp.org/content/22/9/1813.full.pdf+html) and peak count: `*.FRiP_mqc.tsv` and `*_peaks.count_mqc.tsv`.

    *Plots*:  
    [MultiQC - HOMER annotatePeaks peak to genomic feature ratio plot](images/mqc_annotatePeaks_feature_percentage_plot.png)  
    [MultiQC - MACS2 total peak count plot](images/mqc_macs2_peak_count_plot.png)  
    [MultiQC - MACS2 peaks FRiP score plot](images/mqc_frip_score_plot.png)

5. **Create consensus set of peaks**

    *Software*:  
    [BEDTools](https://github.com/arq5x/bedtools2/)

    *Description*:  
    TODO.

    *Output directories*:
    * `bwa/replicate/macs2/merged/`  
      Consensus peak-set across all samples in BED format: `*.bed`.
      HOMER peak-to-gene annotation file for consensus peak-set: `*.annotatePeaks.txt`.   
      Consensus peak-set across all samples in SAF format: `*.saf`. Required by featureCounts.  
      Spreadsheet representation of merged peak set across samples **with** gene annotation columns: `*.boolean.annotatePeaks.txt`. The columns from individual peak files are included in this file along with the ability to filter peaks based on their presence or absence in multiple replicates/conditions.  
      Spreadsheet representation of merged peak set across samples **without** gene annotation columns: `*.boolean.txt`. Use file above for downstream analysis.  
      [UpSetR](https://cran.r-project.org/web/packages/UpSetR/README.html) files to illustrate peak intersection: `*.boolean.intersect.plot.pdf` and `*.boolean.intersect.txt`.  

    *Plots*:  
    [R - UpSetR peak intersection plot](images/mqc_upsetr_intersect_plot.png)

6. **Read counting relative to consensus set of peaks**

    *Software*:  
    [featureCounts](http://bioinf.wehi.edu.au/featureCounts/)

    *Description*:  
    TODO.

    *Output directories*:
    * `bwa/replicate/macs2/merged/deseq2/`  

    *Plots*:  
    [MultiQC - featureCounts consensus peak read assignment plot](images/mqc_featureCounts_assignment_plot.png)

7. **Differential binding analysis, PCA and clustering**

    *Software*:  
    [R](https://www.r-project.org/), [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)

    *Description*:  
    TODO.

    *Output directories*:
    * `bwa/replicate/macs2/merged/deseq2/`  
      Differential binding results across all merged peaks and all comparisons: `*.results.txt`.  
      Plots for PCA, hierarchical clustering,and DESeq2 dispersion estimates and variance stabilizing transformation: `*.plots.pdf`  
      Log file with information for number of genes differentially bound at different FDR and fold-change thresholds for each comparison: `*log`.  
      R data file containing `dds` and `rld` objects generated by DESeq2: `*.dds.rld.RData`.  
      R sessionInfo log file containing information about R, the OS and attached or loaded packages: `R_sessionInfo.log`.  

    * `bwa/replicate/macs2/merged/<COMPARISON>/`  
      Spreadsheet containing comparison-specific DESeq2 output for differential binding results across all peaks: '`*.results.txt`'  
      Subset of above file for peaks that pass FDR <= 0.01 (`*FDR0.01.results.txt`), FDR <= 0.01 and fold-change >= 2 (`*FDR0.01.FC2.results.txt`), FDR <= 0.05 (`*FDR0.05.results.txt`) and FDR <= 0.0The main difference is that multiple libraries sequenced from the same sample will be merged at the replicate-level whereas all the replicates associated with an experimental condition will be merged at the sample-level.5 and fold-change >= 2 (`*FDR0.05.FC2.results.txt`).  
      BED files for peaks that pass FDR <= 0.01 (`*FDR0.01.results.bed`), FDR <= 0.01 and fold-change >= 2 (`*FDR0.01.FC2.results.bed`), FDR <= 0.05 (`*FDR0.05.results.bed`) and FDR <= 0.05 and fold-change >= 2 (`*FDR0.05.FC2.results.bed`).
      MA, Volcano, clustering and scatterplots at FDR <= 0.01 and FDR <= 0.05: `*deseq2.plots.pdf`.  

    * `bwa/replicate/macs2/merged/sizeFactors/`  
      Files containing DESeq2 sizeFactors per sample: `*.txt` and `*.RData`.

    *Plots*:  
    [MultiQC - DESeq2 PCA plot](images/mqc_deseq2_pca_plot.png)
    [MultiQC - DESeq2 sample similarity plot](images/mqc_deseq2_sample_similarity_plot.png)

## Sample-level analysis

`--skipMergeBySample`
The library-level alignments associated with all of the replicates from the same experimental condition are also merged at the sample-level. This can be useful to increase the coverage for peak-calling and for other analyses that require high sequencing depth such as [motif footprinting](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3959825/).

The analysis steps and directory structure for `bwa/replicate/` and `bwa/sample/` are almost identical.

>NB: Replicate-level alignments will be used for read counting relative to the consensus sample-level peakset. This is the only way in which differential analysis can be performed at the sample-level.

## Aggregate analysis

The following directories will be created in the output directory after the pipeline has finished. All paths are relative to the top-level results directory.

1. **Present QC at the raw read, alignment and peak-level**

    *Software*:  
    [MultiQC](http://multiqc.info/)

    *Description*:  
    MultiQC is a visualisation tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available within the report data directory.  

    The pipeline has special steps which allow the software versions used to be reported in the MultiQC output for future traceability.

    *Output directories*:
    * `multiqc/`  
      `Project_multiqc_report.html` - a standalone HTML file that can be viewed in your web browser.
      `Project_multiqc_data/` - directory containing parsed statistics from the different tools used in the pipeline.

    Results generated by MultiQC collate pipeline QC from FastQC, TrimGalore, samtools flagstat, samtools idxstats, picard CollectMultipleMetrics, picard MarkDuplicates, featureCounts. The default [multiqc config file](../assets/multiqc_config.yaml) also contains the provision for loading custom-content to report peak counts, FriP scores, peak to gene annnotation proportions, sample-similarity heatmaps and PCA plots.  

    For more information about how to use MultiQC reports, see http://multiqc.info

2. **Create IGV session file**

    *Software*:  
    [IGV](https://software.broadinstitute.org/software/igv/)

    *Description*:  
    bigWig tracks, peaks and differential sites for data visualisation.

    *Output directories*:
    * `igv/`  
      An IGV session file called `igv_session.xml` will be created at the end of the pipeline. This avoids having to load all the data individually into IGV for visualisation.  

    Once installed, open IGV, go to `File > Open Session` and select the `igv_session.xml` file for loading. File paths in the IGV session file will be set as absolute paths to the directory containing the results. If you prefer to load the data over the web you can just replace the relevant portion of the file path with a link in the session file. The path to the genome fasta file provided to the pipeline will be set as the genome for the IGV session. If you prefer to use an in-built genome provided by IGV just change the file path to the name of the IGV genome e.g. mm10 or hg19.

## Other results

The following directories will be created in the output directory after the pipeline has finished. All paths are relative to the top-level results directory.

1. **Reference genome files**

    *Software*:  
    [BWA](https://sourceforge.net/projects/bio-bwa/files/), [BEDTools](https://github.com/arq5x/bedtools2/), [SAMtools](https://sourceforge.net/projects/samtools/files/samtools/)

    *Description*:    
    TODO.

    *Output directories*:
    * `reference_genome/`  
      If the `--saveReference` parameter is provided then genome-specific files and alignment indices generated by the pipeline will be saved in this directory.

2. **Pipeline information**

    *Software*:  
    [Nextflow!](https://www.nextflow.io/)

    *Description*:  
    Nextflow provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to trouble-shoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

    *Output directories*:
    * `pipeline_info/`  
      Default reports generated by the pipeline are `atacseq_report.html`, `atacseq_timeline.html`, `atacseq_trace.txt` and `atacseq_dag.dot`. See [Nextflow Tracing & visualisation](https://www.nextflow.io/docs/latest/tracing.html).  

    * `Documentation/`  
      Additional reports and documentation generated by the pipeline i.e. `pipeline_report.html`, `pipeline_report.txt`, `results_description.html`.

<!---
The main difference is that multiple libraries sequenced from the same sample will be merged at the replicate-level whereas all the replicates associated with an experimental condition will be merged at the sample-level.
-->
